File: 2179-most-beautiful-item-for-each-query/most-beautiful-item-for-each-query.cpp
class Solution {
public:
    vector<int> maximumBeauty(vector<vector<int>>& items, vector<int>& queries) {
        sort(items.begin(), items.end());
        vector<pair<int, int>> queries2;
        for (int i = 0; i < queries.size(); i++){
            pair<int, int> temp(queries[i], i);
            queries2.push_back(temp);
        }
        sort(queries2.begin(), queries2.end());
        vector<int> result(queries.size(), 0);
        int j = 0;
        int curr_max = 0;
        for (pair<int, int>i : queries2){
            while (j < items.size() && i.first >= items[j][0]){
                curr_max = max(curr_max, items[j][1]);
                j += 1;
            }
            result[i.second] = curr_max;
        }
        
        
        return result;
    }
};

File: 2179-most-beautiful-item-for-each-query/most-beautiful-item-for-each-query.py
class Solution:
    def maximumBeauty(self, items: List[List[int]], queries: List[int]) -> List[int]:
        items.sort()
        queries = sorted([(j, i) for i, j in enumerate(queries)])
        result = [0] * len(queries)
        max_curr = 0
        j = 0
        for q, i in queries:
            while j < len(items) and q >= items[j][0]:
                max_curr = max(max_curr, items[j][1])
                j += 1
            result[i] = max_curr
        return result
        # cache = defaultdict(int)
        # for i, j in items:
        #     cache[i] = max(cache[i], j)
        
        # keys = sorted(cache.keys())
        # curr_max = float("-inf")
        # for i in keys:
        #     if cache[i] > curr_max:
        #         curr_max = cache[i]
        #     else:
        #         cache[i] = curr_max

        # result = []
        # for i in queries:
        #     curr_index = 0
        #     while 


        # return result